### PR TITLE
Add get it

### PR DESCRIPTION
**Pull request by Diogo00dev**
# Short Description
Added get_it as a dependency injection tool
Created a basic UI as ground work

# What changes in the code?
Added get_it
Created Dio http client
Added "Country" query to the topHeadlines request (used "us" as default)
Added some nullable types in the Article class
Changed some ArticlesState types to not be privet, to be able to get the type in unit tests
Added a simple UI as a starting point

# Did you test?
No, how to test get_it implementation ? 🤔

### CODE REVIEW

File: flutter/news/lib/dependencies_injection.dart [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+Future<void> init() async {
+  getIt.registerFactory(() => ArticlesCubit(repository: getIt()));
+
+  getIt.registerLazySingleton(
+    () =>
+        ArticlesRepository(localDataSource: getIt(), remoteDataSource: getIt()),
+  );
+
+  getIt.registerLazySingleton(() => ArticlesLocalDataSource(getIt()));
```
Comment (juankysoriano) [inline flutter/news/lib/dependencies_injection.dart]: perhaps here the named arguments `MyStuff({required dependency, required dependency2})` could be useful to understand by reading this class what are the dependencies of things like the `ArticlesLocalDataSource` (etc).

Comment (juankysoriano) [reply flutter/news/lib/dependencies_injection.dart]: We should revisit this classes and make them have named arguments for the construction

File: flutter/news/lib/dependencies_injection.dart [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+  getIt.registerLazySingleton(() => ArticlesRemoteDataSource(client: getIt()));
+
+  getIt.registerLazySingleton(() => const JsonCodec());
+
+  getIt.registerLazySingleton(() => NewsApiClient(getIt()));
+
+  getIt.registerLazySingleton(() => createDio());
```
Comment (juankysoriano) [inline flutter/news/lib/dependencies_injection.dart]: I would like you to reflect about something.

`createDio` creates a DIO having on the properties the headers for the `NEWS_API_KEY`. The question is, could we use this DIO `ApiClient`s other than the `NewsApiClient`? Perhaps no, because of that header not being required for others. 

With that thought in place, it is possible to imagine an scenario where we have `createNewsApiClientDio()` and `createOtherApiClientDio()`. 

Now the question is, how can we distinguish which one is used when creating the NewsApiClient?

File: flutter/news/lib/core/http_client.dart [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import 'package:dio/dio.dart';
+import 'package:dio_logging_interceptor/dio_logging_interceptor.dart';
+import 'package:flutter_dotenv/flutter_dotenv.dart';
+
+Dio createDio() {
```
Comment (juankysoriano) [inline flutter/news/lib/core/http_client.dart]: Given that this `createDio` is only going to be usable with `NewsApiClient`, does it make sense to have it in a separate file? Perhaps defining it in `news_api_client.dart` makes sense

File: flutter/news/lib/dependencies_injection.dart (hunk 2 above)
Comment (juankysoriano) [reply flutter/news/lib/dependencies_injection.dart]: The same applies when we create multiple singletons of the same type, how does `get-it` distinguish?

File: flutter/news/lib/core/http_client.dart (hunk 3 above)
Comment (juankysoriano) [reply flutter/news/lib/core/http_client.dart]: Perhaps something like this could make sense, because the dio created here makes sense only for `NewsApiClient`, could we use...

```

@RestApi(baseUrl: "https://newsapi.org")
abstract class NewsApiClient {
  factory NewsApiClient(Dio dio, {String baseUrl}) = _NewsApiClient;
  factory NewsApiClient.create() => NewsApiClient(_createDio());

  @GET("/v2/top-headlines")
  Future<BaseNewsResponse> topHeadLines(@Query("country") String country);
}

Dio _createDio() {
  final dio = Dio();
  dio.options.headers['X-Api-Key'] = dotenv.env['NEWS_API_KEY'];
  dio.interceptors.add(
    DioLoggingInterceptor(
      level: Level.body,
      compact: false,
    ),
  );
  return dio;
}
```

Comment (juankysoriano) [reply flutter/news/lib/core/http_client.dart]: Then on the `get-it` implementation, we create the `NewsApiClient` with `NewsApiClient.create()` while keeping the other constructor, that could be useful for tests

File: flutter/news/lib/features/frontpage/data/datasource/articles_remote_data_source.dart [hunk 4]
```diff
@@ -9,10 +9,10 @@ class ArticlesRemoteDataSource {
 
   Future<Result<List<Article>>> topHeadLines() async {
     var result = await client
-        .topHeadLines()
+        .topHeadLines("us")
```
Comment (juankysoriano) [inline flutter/news/lib/features/frontpage/data/datasource/articles_remote_data_source.dart]: I think this could read very well with a named argument for the `topHeadlines` method. `topHeadlines(country: "us")`

File: flutter/news/lib/features/frontpage/presentation/bloc/articles_cubit.dart [hunk 5]
```diff
@@ -30,7 +30,7 @@ class ArticlesCubit extends Cubit<ArticlesState> {
           ),
         )
       },
-      failure: (failure) => emit(const ArticlesState.error()),
+      failure: (failure) => emit(ArticlesState.error(error: failure.toString())),
```
Comment (juankysoriano) [inline flutter/news/lib/features/frontpage/presentation/bloc/articles_cubit.dart]: What could we do here so that we can get the message by doing `failure.message`? Perhaps `InternalFailure` should contain a `message` field? It could make sense because every class implementing it at the moment has it.

For example:
```
abstract class InternalFailure {
  final String message;

  InternalFailure(this.message);
}

@freezed
abstract class ServerFailure extends InternalFailure with _$ServerFailure {
  const factory ServerFailure({required String message}) = _ServerFailure;
}

@freezed
abstract class CacheFailure extends InternalFailure with _$CacheFailure {
  const factory CacheFailure({required String message}) = _CacheFailure;
}

```

File: flutter/news/lib/features/frontpage/presentation/bloc/top_headlines_state.dart [hunk 6]
```diff
@@ -5,10 +5,15 @@ part 'top_headlines_state.freezed.dart';
 
 @freezed
 class ArticlesState with _$ArticlesState {
-  const factory ArticlesState.initial() = _Initial;
-  const factory ArticlesState.loading() = _Loading;
+  const factory ArticlesState.initial() = Initial;
```
Comment (juankysoriano) [inline flutter/news/lib/features/frontpage/presentation/bloc/top_headlines_state.dart]: Where do you need this as `public`?

I don't think there is a case where we need `_Loading`, `_Initial`, `_Error` to be public, right? it should be enough if we deal with `state.when()` when having to see which type is which one

File: flutter/news/lib/features/frontpage/presentation/widgets/loading_widget.dart [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+  @override
+  Widget build(BuildContext context) {
+    return SizedBox(
+      height: MediaQuery.of(context).size.height / 3,
+      child: const Center(
+        child: CircularProgressIndicator(),
+      ),
+    );
+  }
+}
```
Comment (juankysoriano) [inline flutter/news/lib/features/frontpage/presentation/widgets/loading_widget.dart]: needed empty line

File: flutter/news/lib/main.dart [hunk 8]
```diff
@@ -1,117 +1,23 @@
 import 'package:flutter/material.dart';
 import 'package:flutter_dotenv/flutter_dotenv.dart';
+import 'package:news/features/frontpage/presentation/screens/front_page.dart';
+
+import 'dependencies_injection.dart' as di;
 
 void main() async {
-  await dotenv.load(fileName: ".secrets-env");
+  await dotenv.load(fileName: ".secrets_env");
```
Comment (juankysoriano) [inline flutter/news/lib/main.dart:8]: :+1: :100:

File: flutter/news/lib/main.dart [hunk 9]
```diff
@@ -1,117 +1,23 @@
...
-        tooltip: 'Increment',
-        child: const Icon(Icons.add),
-      ), // This trailing comma makes auto-formatting nicer for build methods.
+    return const MaterialApp(
+      title: 'asdsd',
+      home: FrontPage(),
     );
   }
-}
+}
```
Comment (juankysoriano) [inline flutter/news/lib/main.dart]: needs new line

File: flutter/news/test/features/frontpage/presentation/bloc/articles_cubit_test.dart [hunk 10]
```diff
@@ -56,10 +56,7 @@ void main() {
       return ArticlesCubit(repository: repository);
     },
     act: (cubit) => cubit.getTopHeadlines(),
-    expect: () => <ArticlesState>[
-      const ArticlesState.loading(),
-      const ArticlesState.error(),
-    ],
+    expect: () => [isA<Loading>(), isA<Error>()],
```
Comment (juankysoriano) [inline flutter/news/test/features/frontpage/presentation/bloc/articles_cubit_test.dart]: Why this change? I think previously was good enough, also it removes the need for making public `_Loading`, `_Error`, etc

**Comment by juankysoriano** [comment]
:100: Great stuff, let's pair tomorrow @Diogo00dev so I can totally `get-it`  :drum: :drum: :sweat_smile:

**Comment by juankysoriano** [comment]
I have a couple of yummy PRs to be opened after this one is merged

File: flutter/news/test/features/frontpage/presentation/bloc/articles_cubit_test.dart (hunk 10 above)
Comment (Diogo00dev) [reply flutter/news/test/features/frontpage/presentation/bloc/articles_cubit_test.dart]: I changed here because when i added a message to the error i had to had the matching message here, and i think we only need to make sure the state is of the correct Type

File: flutter/news/lib/features/frontpage/presentation/bloc/top_headlines_state.dart (hunk 6 above)
Comment (Diogo00dev) [reply flutter/news/lib/features/frontpage/presentation/bloc/top_headlines_state.dart]: When i added a message to the Error one, i had to change the way i did the test, in the test i am now only checking for the Type (isA<Error>()) so i needed it to be public, and i thought they all should be for the same reason

File: flutter/news/lib/dependencies_injection.dart (hunk 1 above)
Comment (Diogo00dev) [reply flutter/news/lib/dependencies_injection.dart]: updated

File: flutter/news/lib/core/http_client.dart (hunk 3 above)
Comment (Diogo00dev) [reply flutter/news/lib/core/http_client.dart]: updated

File: flutter/news/lib/features/frontpage/presentation/bloc/articles_cubit.dart (hunk 5 above)
Comment (Diogo00dev) [reply flutter/news/lib/features/frontpage/presentation/bloc/articles_cubit.dart]: updated

File: flutter/news/lib/features/frontpage/presentation/widgets/loading_widget.dart (hunk 7 above)
Comment (Diogo00dev) [reply flutter/news/lib/features/frontpage/presentation/widgets/loading_widget.dart]: updated

File: flutter/news/lib/main.dart (hunk 9 above)
Comment (Diogo00dev) [reply flutter/news/lib/main.dart]: updated

**Comment by juankysoriano** [comment]
LGTM :100:
